### PR TITLE
Make serialization work recursively

### LINKED ISSUE #47: Recursive serialization
For example,

```
fac =
| 0 -> 1
| n -> n * fac (n - 1)
```

won't serialize because it results in a closure that references itself.

### PR DESCRIPTION
**Pull request by tekknolagi**
Use `Ref`s to get around recursion in JSON objects (which have no pointers).

### CODE REVIEW

**Comment by tekknolagi** [comment]
Still need to make deserialize work with serialize (refs)

File: scrapscript.py [hunk 1]
```diff
@@ -1359,8 +1292,149 @@ def bdecode(msg: str) -> object:
...
+            link = self.objs[idx]
+            assert isinstance(link, Link)
+            link.link(obj)
+            self.objs[idx] = obj
+            self.link(idx)
+        return self.objs[0]
+
+    # TODO(max): Figure out a way to link without having two entire functions
+    # that have knowledge about each object type
+    def link(self, idx: int) -> None:
```
Comment (tekknolagi) [inline scrapscript.py:1408]: @gregorybchris have any ideas on how to do this? i am using one level of indirection but it's a little gross

Comment (tekknolagi) [reply scrapscript.py:1408]: in particular the gross thing is having two tree traversals. if we can get it to one traversal, that would be fine imo

Comment (gregorybchris) [reply scrapscript.py:1408]: No good ideas for you...

**Comment by tekknolagi** [comment]
Superseded by #169

**Pull request closed without merging**